### PR TITLE
JIT: add extra check to struct of struct of x promotion

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Avoid promoting structs that contain struct fields that themselves
wrap single simple fields, if those single simple fields are smaller
than their enclosing struct.

Otherwise we run the risk of losing track of the "extra" bytes in the
inner struct.

Addresses #19149.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@briansull PTAL
cc @dotnet/jit-contrib 

@mgravell if you have bigger or other test cases around, feel free to give this a try or send me pointers.

No diffs for pmi across x64 fx.

Still need to add test case and look at whether there's something that we could do to flag the potential off the end access (so we could keep promotion but make it dependent, perhaps?).

File: src/jit/lclvars.cpp [hunk 1]
```diff
@@ -1656,6 +1656,18 @@ void Compiler::lvaCanPromoteStructType(CORINFO_CLASS_HANDLE    typeHnd,
...
+                    return;
+                }
+
+                // Insist this wrapped field occupy all of its parent storage.
+                unsigned innerStructSize = info.compCompHnd->getClassSize(pFieldInfo->fldTypeHnd);
+
+                if (fieldSize != innerStructSize)
+                {
+                    JITDUMP("Promotion blocked: struct contains struct field with one field,"
+                            "but that field is not the same size as its parent.");
```
Comment (erozenfeld) [inline src/jit/lclvars.cpp]: Missing space before 'but'.

**Comment by AndyAyersMS** [comment]
Reran jit-diffs across all pri0 tests... including the new test. It's the only case with diffs.
```
PMI Diffs for System.Private.CoreLib.dll, framework assemblies, assemblies in d:\repos\coreclr2\bin\tests\Windows_NT.x64.Checked for x64 default jit
Summary:
(Lower is better)
Total bytes of diff: -29 (0.00% of base)
    diff is an improvement.
Top file improvements by size (bytes):
         -29 : JIT\Regression\JitBlue\GitHub_19149\GitHub_19149\GitHub_19149.dasm (-0.44% of base)
1 total files with size differences (1 improved, 0 regressed), 3131 unchanged.
Top method regressions by size (bytes):
          53 : JIT\Regression\JitBlue\GitHub_19149\GitHub_19149\GitHub_19149.dasm - CommandBytes:Equals(ref):bool:this
          14 : JIT\Regression\JitBlue\GitHub_19149\GitHub_19149\GitHub_19149.dasm - Program:<Main>g__HuntFor|0_1(ref,byref):bool
Top method improvements by size (bytes):
         -64 : JIT\Regression\JitBlue\GitHub_19149\GitHub_19149\GitHub_19149.dasm - <>c:<Main>b__0_4(struct):ref:this
         -32 : JIT\Regression\JitBlue\GitHub_19149\GitHub_19149\GitHub_19149.dasm - Program:<Main>g__Add|0_0(ref,byref)
4 total methods with size differences (2 improved, 2 regressed), 363933 unchanged.
```

**Comment by AndyAyersMS** [comment]
All these new failures seem unrelated (AVX_ro and infratructure). The first run had the same code, the new commit fixed a a comment and added a test case.

**Comment by tannergooding** [comment]
I am tracking the `AVX_ro` issue and actively investigating.

CC. @dotnet/dnceng on the Out of Space issue that occurred for the Ubuntu run.

**Comment by AndyAyersMS** [comment]
Am going to ignore these recent failures.

**Pull request merged**